model/vector_quantizer.py

```diff
@@ -0,0 +1,119 @@
+        else:
+            encoding_distances = None
+
+        # Compute distances between embedding vectors
+        if not self.training and compute_distances_if_possible:
+            _embedding_distances = [
+                torch.dist(items[0], items[1], 2).to(self._device) for items in combinations(self._embedding.weight, r=2)
+            ]
+            embedding_distances = torch.tensor(_embedding_distances).to(self._device)
+        else:
+            embedding_distances = None
+
+        # Sample nearest embedding
+        if not self.training and compute_distances_if_possible:
+            _frames_vs_embedding_distances = [
+                torch.dist(items[0], items[1], 2).to(self._device)
+                for items in product(flat_input, self._embedding.weight.detach())
+            ]
+            frames_vs_embedding_distances = (
+                torch.tensor(_frames_vs_embedding_distances).to(self._device).view(batch_size, time, -1)
+            )
+        else:
+            frames_vs_embedding_distances = None
+
+        # Quantize and unflatten
+        quantized = torch.matmul(encodings, self._embedding.weight).view(input_shape)
+        # TODO: Check if the more readable self._embedding.weight.index_select(dim=1, index=encoding_indices) works better
+
+        concatenated_quantized = (
+            self._embedding.weight[torch.argmin(distances, dim=1).detach().cpu()]
+            if not self.training or record_codebook_stats
+            else None
+        )
+
+        # Losses
+        e_latent_loss = torch.mean((quantized.detach() - inputs) ** 2)
+        q_latent_loss = torch.mean((quantized - inputs.detach()) ** 2)
+        commitment_loss = self._commitment_cost * e_latent_loss
+        vq_loss = q_latent_loss + commitment_loss
+
+        quantized = inputs + (quantized - inputs).detach()  # Trick to prevent backpropagation of quantized
+        avg_probs = torch.mean(encodings, dim=0)
+
+        """
+        The perplexity a useful value to track during training.
+        It indicates how many codes are 'active' on average.
+        """
+        perplexity = torch.exp(-torch.sum(avg_probs * torch.log(avg_probs + 1e-10)))  # Exponential entropy
+
+        # Convert quantized from BHWC -> BCHW
+        return (
+            vq_loss,
+            quantized.permute(2, 0, 1).contiguous(),
+            perplexity,
+            encodings.view(batch_size, time, -1),
+            distances.view(batch_size, time, -1),
+            encoding_indices,
+            {
+                "e_latent_loss": e_latent_loss.item(),
+                "q_latent_loss": q_latent_loss.item(),
+                "commitment_loss": commitment_loss.item(),
+                "vq_loss": vq_loss.item(),
+            },
+            encoding_distances,
+            embedding_distances,
+            frames_vs_embedding_distances,
+            concatenated_quantized,
+        )
```

Comment:
Can we use a dict or maybe a dataclass for this? 20 element tuples are untraceable

Reply:
changed to dict